src/windows.rs

```diff
@@ -107,3 +110,172 @@ impl PutAddress for ProcessHandle {
+/// of this process, returning the base address of it.
+#[allow(clippy::clippy::cast_possible_truncation)] // for size_of as u32
+impl ModuleInfo for Pid {
+    fn get_module_base(&self, name: &str) -> std::io::Result<usize> {
+        // taken from https://stackoverflow.com/questions/41552466/how-do-i-get-the-physical-baseaddress-of-an-dll-used-in-a-process
+        let mut module_entry = tlhelp32::MODULEENTRY32 {
+            dwSize: 0,
+            th32ModuleID: 0,
+            th32ProcessID: 0,
+            GlblcntUsage: 0,
+            ProccntUsage: 0,
+            modBaseAddr: std::ptr::null_mut(), // yikes
+            modBaseSize: 0,
+            hModule: std::ptr::null_mut(), // yikes
+            szModule: [0; tlhelp32::MAX_MODULE_NAME32 + 1],
+            szExePath: [0; winapi::shared::minwindef::MAX_PATH],
+        };
+
+        unsafe {
+            module_entry.dwSize = std::mem::size_of::<tlhelp32::MODULEENTRY32>() as u32;
+
+            let snapshot = tlhelp32::CreateToolhelp32Snapshot(
+                tlhelp32::TH32CS_SNAPMODULE | tlhelp32::TH32CS_SNAPMODULE32,
+                *self,
+            );
+            if snapshot == INVALID_HANDLE_VALUE {
+                return Err(std::io::Error::last_os_error());
+            }
+
+            if tlhelp32::Module32First(snapshot, &mut module_entry) == minwindef::TRUE {
+                // makeshift do-while
+                loop {
+                    if utf8_to_string(&module_entry.szModule) == name {
+                        let addr = module_entry.modBaseAddr as usize;
+                        if CloseHandle(snapshot) == minwindef::FALSE {
+                            panic!("Could not close handle")
+                        };
+                        return Ok(addr);
+                    }
+
+                    if tlhelp32::Module32Next(snapshot, &mut module_entry) == minwindef::FALSE {
+                        break;
+                    }
+                }
+            }
+
+            // We searched everything, nothing found
+            if CloseHandle(snapshot) == minwindef::FALSE {
+                panic!("Could not close handle")
+            };
+            Err(std::io::Error::new(
+                std::io::ErrorKind::NotFound,
+                format!(
+                    "Process PID#{} contains no module named \"{}\".",
+                    *self, name
+                ),
+            ))
+        }
+    }
+}
+
+/// Get the process ID of some process by name. For example,"MyGame.exe".
+/// If you want to get the PID of your own process, use `std::process:id() as Pid` instead.
+///
+/// # Errors
+/// If no process exists of that name, returns an `std::io::Error` with kind `std::io::ErrorKind::NotFound`.
+/// If something went very wrong with the windows API, returns last OS error.
+#[allow(clippy::cast_possible_truncation)]
+pub fn get_pid(process_name: &str) -> std::io::Result<Pid> {
```

Comment:
I previously hadn't added a `get_pid` function as there is the question of "what should you do if there is more than one process with the same name?" - which I think is something that should be left up to the library consumer instead of the library.

For example, if you are trying to look at and modify a bunch of calculator or notepad instances, then you might want to get all of the relevant PIDs or some subset, but if you're modifying a game then you only have to worry about a single process.

Do you think there'd be a good way to design a `get_pid` function around the possibility of multiple processes?

Reply:
I suppose I could return an iterator over all processes instead, which returns small structs with just the basics, such as process name, pid, maybe Architecture if it's detectable. Then users can `.filter` those as needed. Though a non-goal should be a wrapper around the entirety of winapi haha. Just, whatever is useful for messing with games, CheatEngine, etc, I suppose? Hence why having a `get_pid` in the first place, it's just so convenient.

I'd do a similar iterator thing for `LibraryInfo`.